### PR TITLE
Bot API 7.0 and 7.1

### PR DESCRIPTION
**Pull request by karb0f0s**
Implements #1334

### CODE REVIEW

File: src/Telegram.Bot/Types/ReactionType.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+[JsonObject(MemberSerialization.OptIn, NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
+public class ReactionTypeEmoji : ReactionType
+{
+    /// <summary>
+    /// Type of the reaction, always "emoji"
+    /// </summary>
+    public override string Type => "emoji";
+
+    /// <summary>
+    /// Reaction emoji. Currently, it can be one of "👍", "👎", "❤", "🔥", "🥰", "👏", "😁",
```
Comment (tuscen) [inline src/Telegram.Bot/Types/ReactionType.cs:35]: I think we should add a static class with const fields for each emoji to make it more discoverable. You can look names for each emoji here http://emojipedia.org
```csharp
public static class KnownReactionTypeEmoji
{
    public const string ThumbsUp = "👍";
    // etc
}
```

File: src/Telegram.Bot/Types/Update.cs [hunk 2]
```diff
@@ -119,20 +136,22 @@ public class Update
...
+        { EditedChannelPost: { } }    => UpdateType.EditedChannelPost,
+        { ShippingQuery: { } }        => UpdateType.ShippingQuery,
+        { PreCheckoutQuery: { } }     => UpdateType.PreCheckoutQuery,
+        { Poll: { } }                 => UpdateType.Poll,
+        { PollAnswer: { } }           => UpdateType.PollAnswer,
+        { MyChatMember: { } }         => UpdateType.MyChatMember,
+        { ChatMember: { } }           => UpdateType.ChatMember,
+        { ChatJoinRequest: { } }      => UpdateType.ChatJoinRequest,
+        { MessageReaction: { } }      => UpdateType.MessageReaction,
+        { MessageReactionCount: { } } => UpdateType.MessageReactionCount,
```
Comment (tuscen) [inline src/Telegram.Bot/Types/Update.cs:154]: Can you replace `{ }` to `not null` in each arm like so `{ Message: not null }`? It reads more explicitly like that.

File: src/Telegram.Bot/Types/ReactionType.cs (hunk 1 above)
Comment (karb0f0s) [reply src/Telegram.Bot/Types/ReactionType.cs:35]: I'll die on that hill :(

File: src/Telegram.Bot/Requests/Available methods/Messages/SetMessageReactionRequest.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    [JsonProperty(Required = Required.Always)]
+    public int MessageId { get; }
+
+    /// <summary>
+    /// New list of reaction types to set on the message. Currently, as non-premium users, bots can
+    /// set up to one reaction per message. A custom emoji reaction can be used if it is either
+    /// already present on the message or explicitly allowed by chat administrators.
+    /// </summary>
+    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+    public ReactionType[]? Reaction { get; set; }
```
Comment (tuscen) [inline src/Telegram.Bot/Requests/Available methods/Messages/SetMessageReactionRequest.cs:33]: Change it to `IEnumerable<ReactionType>`, please. We use it everywhere in the inputs

File: src/Telegram.Bot/TelegramBotClientExtensions.ApiMethods.cs [hunk 4]
```diff
@@ -1804,6 +1804,49 @@ await botClient.ThrowIfNull()
...
+    /// Pass <see langword="true"/> to set the reaction with a big animation
+    /// </param>
+    /// <param name="cancellationToken">
+    /// A cancellation token that can be used by other objects or threads to receive notice of cancellation
+    /// </param>
+    public static async Task SetMessageReactionAsync(
+        this ITelegramBotClient botClient,
+        ChatId chatId,
+        int messageId,
+        ReactionType[]? reaction,
```
Comment (tuscen) [inline src/Telegram.Bot/TelegramBotClientExtensions.ApiMethods.cs:1836]: Here's also `IEnumerable<ReactionType>`

File: src/Telegram.Bot/Types/ReactionType.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+using Telegram.Bot.Converters;
+
+namespace Telegram.Bot.Types;
+
+/// <summary>
+/// This object describes the type of a reaction. Currently, it can be one of
+/// <list type="bullet">
+/// <item><description>ReactionTypeEmoji</description></item>
+/// <item><description>ReactionTypeCustomEmoji</description></item>
```
Comment (tuscen) [inline src/Telegram.Bot/Types/ReactionType.cs:9]:
```suggestion
/// <item><description><see cref="ReactionTypeEmoji"/></description></item>
/// <item><description><see cref="ReactionTypeCustomEmoji"/></description></item>
```

File: src/Telegram.Bot/Types/ReactionType.cs (hunk 1 above)
Comment (karb0f0s) [reply src/Telegram.Bot/Types/ReactionType.cs:35]: addressed

File: src/Telegram.Bot/Extensions/HttpContentExtensions.cs [hunk 6]
```diff
@@ -21,6 +21,7 @@ internal static class HttpContentExtensions
 
         // It will be dispose of after the request is made
 #pragma warning disable CA2000
+        inputFile.Content.Position = 0;
```
Comment (tuscen) [inline src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: Not every stream is rewindable. We need to think of another solution

Comment (tuscen) [reply src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: One idea is to make a buffer stream using `MemoryStream`, though I'm not sure how to deal with disposing of that stream

Comment (karb0f0s) [reply src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: What if we cache file content in Lazy<whatever> field and reuse it on consequent calls?

Comment (karb0f0s) [reply src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: So, I was able to cache content, but it'll require quite a significant change:

```csharp
public ctor()
{
    cachedStreamContent = new(Do);
}

private readonly Lazy<Task<StreamContent?>> cachedStreamContent;

private async Task<StreamContent?> Do()
{
    HttpContent? content =  ...;
    var stream = await content!.ReadAsStreamAsync().ConfigureAwait(false);
    return new StreamContent(stream);
}

public override Task<StreamContent?> HttpContent() => cachedStreamContent.Value;
```

BTW .NET 6+ has synchronous `ReadAsStream()` method.

Comment (karb0f0s) [reply src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: #1339

Comment (karb0f0s) [reply src/Telegram.Bot/Extensions/HttpContentExtensions.cs:24]: #1341

**Comment by smetlov** [comment]
Got the ETA for this one?

**Review by tuscen — DISMISSED** [review]

**Comment by CumpsD** [comment]
Looking forward to be able to use Reactions! :)

**Review by tuscen — APPROVED** [review]

**Comment by CumpsD** [comment]
Any idea when this will be released?

**Comment by tuscen** [comment]
@CumpsD https://github.com/TelegramBots/Telegram.Bot/issues/1375

**Comment by CumpsD** [comment]
> @CumpsD #1375

Thanks! Solved it by downloading the nupkg from releases and adding this to my csproj:

```
    <PropertyGroup>
        <RestoreSources>$(RestoreSources);../assets;https://api.nuget.org/v3/index.json</RestoreSources>
    </PropertyGroup>
```

(nupkg is stored in folder `assets`)

**Pull request merged**